src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+        unsafe {
+            self.reset_overlapped();
+            let buf_ptr = self.read_buf.as_mut_ptr() as LPVOID;
+            let max_read_bytes = buf_cap - buf_len;
+            let ok = kernel32::ReadFile(self.handle,
+                                        buf_ptr.offset(buf_len as isize),
+                                        max_read_bytes as u32,
+                                        &mut bytes_read,
+                                        self.ov_ptr());
+
+            // ReadFile can return TRUE; if it does, an IO completion
+            // packet is still posted to any port, and the OVERLAPPED
+            // structure has the IO operation flagged as complete.
+            if ok == winapi::FALSE {
+                let err = GetLastError();
+                if err == winapi::ERROR_BROKEN_PIPE {
+                    dd!("[$ {:?}:{:?}] BROKEN_PIPE straight from ReadFile", self.iocp, self.handle);
+                    self.closed = true;
+                    return Ok(());
+                }
+
+                if err == winapi::ERROR_IO_PENDING {
+                    self.read_in_progress = true;
+                    return Ok(());
+                }
+
+                Err(WinError::from_system(err, "ReadFile"))
+            } else {
+                self.read_in_progress = true;
+                Ok(())
+            }
+        }
+    }
+
+    fn message_length(&self) -> Option<(usize,usize)> {
+        dd!("[$ {:?}:{:?}] message_length read_buf: {} bytes", self.iocp, self.handle, self.read_buf.len());
+        if self.read_buf.len() < HEADER_SIZE {
+            return None;
+        }
+
+        let buf_header: &[u32] = unsafe { slice::from_raw_parts(self.read_buf.as_ptr() as *const u32, 2) };
+        let data_bytes = buf_header[0] as usize;
+        let oob_bytes = buf_header[1] as usize;
+        let bytes_needed = HEADER_SIZE + data_bytes + oob_bytes;
+        if self.read_buf.len() >= bytes_needed {
+            Some((data_bytes, oob_bytes))
+        } else {
+            None
+        }
+    }
+
+    // Err(false) -> something really failed
+    // Err(true) -> no message
+    // FIXME This is dumb, we should probably make this return Result<Option<(...)>,WinError>
+    // so that we can pass through the error that's lost in map_err below
+    fn get_message(&mut self) -> Result<(Vec<u8>, Vec<OsOpaqueIpcChannel>, Vec<OsIpcSharedMemory>),bool> {
+        let message_lengths = self.message_length();
+        if message_lengths.is_none() {
+            return Err(false)
+        }
+        let (data_bytes, oob_bytes) = message_lengths.unwrap();
+
+        let bytes_needed = HEADER_SIZE + data_bytes as usize + oob_bytes as usize;
+
+        // remove this message's bytes from read_buf, or just take read_buf
+        // if it contains exactly one message
+        let msg_buf = if self.read_buf.len() == bytes_needed {
+            mem::replace(&mut self.read_buf, Vec::with_capacity(READ_BUFFER_SIZE))
```

Comment:
So apparently `drain()` isn't doing the optimisation I had hoped it would...

As for the better performance with small sizes, I doubt it's avoiding the branch that makes a significant difference here. I suspect it's rather reduced allocation that makes it faster. (With `drain()` it only allocates a tiny new `Vec<>` for the actually copied data -- as opposed to needing an entire new `PIPE_BUFFER_SIZE` buffer...)

You could indeed adjust the conditional to always use `drain()` for small sizes: this kind of micro-optimisation is actually justified for once I'd say, when it gains a noticeable speed-up for such a fundamental mechanism...

(I'm now *seriously* considering doing a similar optimisation in the `unix` back-end -- even though caching the buffer will complicate the code, and increase memory usage. Receive buffer allocation makes up a *huge* part of the time spent in small message transfers on my system -- IIRC something like 35%...)